services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/grid/internal/replica/merge/GridCsvImportDaoImplTest.java

```diff
@@ -0,0 +1,214 @@
+		gridIndexManger.truncateAll();
+	}
+	
+	@Test
+	public void testImportAndJoin() throws IOException {
+		
+		writeGridData();
+		
+		GridHeader gridHeader = gridViewManager.readHeader(sessionId, replicaId).orElseThrow();
+		
+		long start = System.currentTimeMillis();
+		
+		Iterator<RowView> rowView = gridViewManager.getQueryIterator(gridHeader, Collections.emptyList());
+		
+		DataStream gridDataStream = new GridDataStream(rowView, columnMapping);
+		
+		importDao.streamToGridTempTable(gridDataStream, columnMapping);
+		
+		System.out.println("Grid import took: " + (System.currentTimeMillis() - start) + "ms");
+		
+		Iterator<Object[]> it = importDao.getGridTempTableIterator();
+		
+		long rowId = 0;
+		Long rowVecId = 22L;
+		
+		while (it.hasNext()) {
+			Object[] row = it.next();
+			// This is the extra column with information from the grid (serialized vector id)
+			String extraData = "[\"[" + replicaId + "," + rowVecId + "]\"]";
+			// Note that the data has been reordered by the upsert key (b,a) and only contains the upsert keys plus the extra data column
+			assertArrayEquals(new Object[] { rowId, rowId % 10, extraData }, row);
+			rowId++;
+			rowVecId+=9;
+		}
+		
+		// Now import the CSV data (contains a subset of the grid data, with additional columns and different order)
+		
+		start = System.currentTimeMillis();
+		
+		CSVReader reader = CSVUtils.createCSVReader(new StringReader(csvContent), null, null);
+		CsvDataStream csvDataStream = new CsvDataStream(reader, columnMapping);
+		
+		try (reader) {
+			importDao.streamToCsvTempTable(csvDataStream, columnMapping);
+		}
+
+		System.out.println("CSV import took: " + (System.currentTimeMillis() - start) + "ms");
+		
+		it = importDao.getCsvTempTableIterator();
+		rowId = 0;
+		
+		while (it.hasNext()) {
+			String expectedCsvJson = "[false,\"bar" + rowId + "\"]";
+			
+			assertArrayEquals(
+				// Note that column e is omitted as not present in the grid and the columns that are not
+				// the upsert key are stored in a JSON array
+				new Object[] { rowId, rowId % 5, expectedCsvJson },
+				it.next()
+			);
+			
+			rowId++;
+		}
+	
+		Iterator<JoinedRow> joinIt = importDao.getJoinedTempTableIterator(columnMapping);
+		
+		rowId = 0;
+		rowVecId = 22L;
+		int notMatchedCount = 0;
+		
+		while (joinIt.hasNext()) {
+			
+			Object[] expectedCsvData = new Object[] { rowId, rowId % 5, false , "bar" + rowId };
+			Object[] expectedGridData = null;
+			
+			if (rowId % 10 < 5 && rowId < gridRowCount) {
+				expectedGridData = new Object[] { "[" + replicaId + "," + rowVecId + "]" };
+			} else {
+				notMatchedCount++;
+			}
+			
+			JoinedRow next = joinIt.next();
+			
+			assertArrayEquals(expectedCsvData, next.getCsvData());
+			assertArrayEquals(expectedGridData, next.getGridData());
+			
+			rowId++;
+			rowVecId+=9;
+		}
+		
+		// The CSV has only 50% of the rows in common with the grid
+		assertEquals(gridRowCount/2 + (csvRowCount - gridRowCount), notMatchedCount);
```

Comment:
Nitpicking...if the test was tweaked so `gridRowCount > csvRowCount`, this could be negative...would this account for it?


```suggestion
		assertEquals(gridRowCount/2 + Math.abs(csvRowCount - gridRowCount), notMatchedCount);
```